barretenberg/cpp/pil/avm/avm_main.pil

```diff
@@ -419,6 +452,42 @@ namespace avm_main(256);
     #[TIMESTAMP_KERNEL]
     sel_op_timestamp * (avm_kernel.kernel_sel - constants.TIMESTAMP_SELECTOR) = 0;
 
+    // OUTPUTS LOOKUPS
+    // Constrain the value of kernel_out_sel to be the correct offset for the operation being performed
+    #[NOTE_HASH_KERNEL_OUTPUT]
+    sel_op_note_hash_exists * (avm_kernel.kernel_out_sel - (avm_kernel.START_NOTE_HASH_EXISTS_WRITE_OFFSET + avm_kernel.note_hash_exist_write_offset)) = 0;
+
+    #[EMIT_NOTE_HASH_KERNEL_OUTPUT]
+    sel_op_emit_note_hash * (avm_kernel.kernel_out_sel - (avm_kernel.START_EMIT_NOTE_HASH_WRITE_OFFSET + avm_kernel.emit_note_hash_write_offset)) = 0;
+
+    #[NULLIFIER_EXISTS_KERNEL_OUTPUT]
+    sel_op_nullifier_exists * (avm_kernel.kernel_out_sel - (avm_kernel.START_NULLIFIER_EXISTS_OFFSET + avm_kernel.nullifier_exists_write_offset)) = 0;
+
+    #[EMIT_NULLIFIER_KERNEL_OUTPUT]
+    sel_op_emit_nullifier * (avm_kernel.kernel_out_sel - (avm_kernel.START_EMIT_NULLIFIER_WRITE_OFFSET + avm_kernel.emit_nullifier_write_offset)) = 0;
+
+    #[L1_TO_L2_MSG_EXISTS_KERNEL_OUTPUT]
+    sel_op_l1_to_l2_msg_exists * (avm_kernel.kernel_out_sel - (avm_kernel.START_L1_TO_L2_MSG_EXISTS_WRITE_OFFSET + avm_kernel.l1_to_l2_msg_exists_write_offset)) = 0;
+
+    #[EMIT_UNENCRYPTED_LOG_KERNEL_OUTPUT]
+    sel_op_emit_unencrypted_log * (avm_kernel.kernel_out_sel - (avm_kernel.START_EMIT_UNENCRYPTED_LOG_WRITE_OFFSET + avm_kernel.emit_unencrypted_log_write_offset)) = 0;
+
+    #[EMIT_L2_TO_L1_MSGS_KERNEL_OUTPUT]
+    sel_op_emit_l2_to_l1_msg * (avm_kernel.kernel_out_sel - (avm_kernel.START_EMIT_L2_TO_l1_MSG + avm_kernel.emit_l2_to_l1_msg_write_offset)) = 0;
+
+    #[SLOAD_KERNEL_OUTPUT]
+    sel_op_sload * (avm_kernel.kernel_out_sel - (avm_kernel.START_SLOAD_WRITE_OFFSET + avm_kernel.sload_write_offset)) = 0;
+
+    #[SSTORE_KERNEL_OUTPUT]
+    sel_op_sstore * (avm_kernel.kernel_out_sel - (avm_kernel.START_SSTORE_WRITE_OFFSET + avm_kernel.sstore_write_offset)) = 0;
+
+    // When we encounter a state writing opcode
+    // We increment the side effect counter by 1
+    KERNEL_OUTPUT_SELECTORS * (avm_kernel.side_effect_counter' - (avm_kernel.side_effect_counter + 1)) = 0;
```

Comment:
Side effect counter will be required to be constrained to the value of the start_side_effect_counter in the public inputs - this can be done with a copy constraint or via a lookup